runtime/src/main/cpp/utf8/core.h

```diff
@@ -187,6 +197,11 @@ namespace internal
 
         code_point = ((code_point << 12) & 0xffff) + ((utf8::internal::mask8(*it) << 6) & 0xfff);
 
+        if (utf8::internal::is_surrogate(code_point))
```

Comment:
Fail as early as possible on invalid sequence to replace each byte with the replacement code point.